magenta-lib/src/main/scala/magenta/artifact/S3Location.scala

```diff
@@ -0,0 +1,131 @@
+  def relativeTo(path: S3Location): String = {
+    key.stripPrefix(path.key).stripPrefix("/")
+  }
+  def fetchContentAsString()(implicit client: AmazonS3) = S3Location.fetchContentAsString(this)
+  def listAll()(implicit client: AmazonS3) = S3Location.listObjects(this)
+}
+
+object S3Location extends Loggable {
+  def listAll(bucket: String)(implicit s3Client: AmazonS3): Seq[S3Object] = listObjects(bucket, None)
+
+  def listObjects(location: S3Location)(implicit s3Client: AmazonS3): Seq[S3Object] =
+    listObjects(location.bucket, Some(location.key))
+
+  val maxKeysInBucketListing = 1000 // AWS won't return more than this, even if you set the parameter to a larger value
+
+  private def listObjects(bucket: String, prefix: Option[String])(implicit s3Client: AmazonS3): Seq[S3Object] = {
+    def request(continuationToken: Option[String]) = new ListObjectsV2Request()
+      .withBucketName(bucket)
+      .withPrefix(prefix.orNull)
+      .withMaxKeys(maxKeysInBucketListing)
+      .withContinuationToken(continuationToken.orNull)
+
+    @tailrec
+    def pageListings(acc: Seq[ListObjectsV2Result], previousListing: ListObjectsV2Result): Seq[ListObjectsV2Result] = {
+      if (!previousListing.isTruncated) {
+        acc
+      } else {
+        val listing = s3Client.listObjectsV2(
+          request(Some(previousListing.getNextContinuationToken))
+        )
+        pageListings(acc :+ listing, listing)
+      }
+    }
+
+    val initialListing = s3Client.listObjectsV2(request(None))
+    for {
+      summaries <- pageListings(Seq(initialListing), initialListing)
+      summary <- summaries.getObjectSummaries.asScala
+    } yield S3Object(summary.getBucketName, summary.getKey, summary.getSize)
+  }
+
+  def fetchContentAsString(location: S3Location)(implicit client:AmazonS3):Option[String] = {
+    Try {
+      Some(client.getObjectAsString(location.bucket, location.key))
+    }.recover {
+      case e: AmazonS3Exception if e.getStatusCode == 404 => None
+    }.get
+  }
+}
+
+case class S3Path(bucket: String, key: String) extends S3Location
+
+object S3Path {
+  def apply(location: S3Location, key: String): S3Path = S3Path(location.bucket, s"${location.key}/$key")
+}
+
+case class S3Object(bucket: String, key: String, size: Long) extends S3Location
+
+object S3Artifact extends Loggable {
+  def apply(build: Build, bucket: String): S3Artifact = {
+    val prefix = buildPrefix(build)
+    S3Artifact(bucket, prefix)
+  }
+
+  def buildPrefix(build: Build): String = {
+    s"${build.projectName}/${build.id}"
+  }
+
+  /*
+  This is essentially the same as the apply method in function, but in this case we actually confirm that the
+  deployObject exists. If it doesn't then we try to convert a legacy artifacts.zip file into the new format.
+   */
+  def withLegacyFallback(build: Build, bucket: String)(implicit client: AmazonS3, reporter: DeployReporter): S3Artifact = {
```

Comment:
I'd expect `withLegacyFallback` to take a function or a `Try` and then recover from the failure case, rather than doing both the happy and sad cases itself.

Also, I'd probably err away from calling it `Legacy`. Maybe something like `ZipBundle`?